app/store/migrations.js

```diff
@@ -47,7 +48,30 @@ export const migrations = {
 			};
 		}
 		return state;
+	},
+	3: state => {
+		const provider = state.engine.backgroundState.NetworkController.provider;
+		const chainId = NetworksChainId[provider.type];
+		// if chainId === '' is a rpc
+		if (chainId) {
+			state.engine.backgroundState.NetworkController.provider = { ...provider, chainId };
+			return state;
+		}
+
+		// If provider is rpc, check if the current network has a valid chainId
+		const chainIdNumber = parseInt(provider.chainId, 10);
+		const isCustomRpcWithInvalidChainId = !isSafeChainId(chainIdNumber);
```

Comment:
`parseInt` can behave weirdly, so I think we should make sure that it only contains numbers, by refactoring like this:
```suggestion
		// If provider is rpc, check if the current network has a valid chainId
		const storedChainId = typeof provider.chainId === 'string' ? provider.chainId : '';
		const isNumericalString = Boolean(storedChainId.match(/^[1-9]\d*$/u))
		const isCustomRpcWithInvalidChainId =  !isNumericalString || !isSafeChainId(parseInt(storedChainId, 10));
```

Reply:
Ref:
```
> parseInt('1x1aaaa', 10)
1
```

Reply:
i think instead of the `typeof` check could we instead do:
```
const storedChainId = provider?.chainId || '';
```
> so I think we should make sure that it only contains numbers

could we do that by coercing with `Number()` instead? If it's possible that might be clearer than using `parseInt()`?

also, for the regex you could get rid of the `Boolean` coercion if you used `.test` I _think_ since that'll return a bool for you out of box:

```js
/^[1-9]\d*$/u.test('19.20')
false
/^[1-9]\d*$/u.test('19')
true
```
@rekmarks lmk what you think

Reply:
Using `test` is a good idea! I updated my snippet to do so.

As for the optional chaining, will `state.engine.backgroundState.NetworkController.provider` ever be `null` or `undefined`? I thought it would always be an object.

I think we should keep the `typeof` check for the `chainId` property, I don't see any harm in doing so. 🤔

Reply:
the optional chaining is more to account for `chainId` ever being `null` or `undefined` here. is there ever a time when `chainId` is _not_ a string?

Reply:
> is there ever a time when `chainId` is _not_ a string?

We decided out of band that, while we don't necessarily have any reason to believe that it could be a non-string value if present, better safe than sorry with migrations!

Reply:
Also, I do agree that we might as well do this:

```
const { provider = {}} = state.engine.backgroundState.NetworkController;
```

Reply:
commiting the change and I literally just copy pasted the previous migration for that 😅